components/authz-service/server/v2/rules_property_test.go

```diff
@@ -82,88 +579,94 @@ func TestCreateRuleProperties(t *testing.T) {
+}
 
-			_, err = cl.CreateRule(ctx, &reqs.rules[0])
-			// Note: this could be very noisy, but it's hard to figure out what went
-			// wrong without the error
-			if err != nil {
-				t.Error(err.Error())
-				return false // bad run
-			}
+func reportErrorAndYieldFalse(t *testing.T, err error) bool {
+	t.Helper()
+	t.Error(err.Error())
+	return false
+}
 
-			r, err := cl.GetRule(ctx, &api.GetRuleReq{Id: reqs.rules[0].Id})
-			if err != nil {
-				t.Error(err.Error())
+func ruleMatches(req interface{}, actual api.ProjectRule) bool {
+	if _, ok := req.(api.CreateRuleReq); ok {
+		ruleReq, _ := req.(api.CreateRuleReq)
+		return len(actual.Conditions) == len(ruleReq.Conditions) &&
+			actual.Id == ruleReq.Id &&
+			actual.Name == ruleReq.Name &&
+			actual.ProjectId == ruleReq.ProjectId
+	} else {
+		ruleReq, _ := req.(api.UpdateRuleReq)
+		return len(actual.Conditions) == len(ruleReq.Conditions) &&
+			actual.Id == ruleReq.Id &&
+			actual.Name == ruleReq.Name &&
+			actual.ProjectId == ruleReq.ProjectId
+	}
+}
+
+func rulesMatch(rules []api.CreateRuleReq, resp *api.ListRulesResp) bool {
+	if len(rules) != len(resp.Rules) {
+		return false
+	}
+	// Cannot assume DB returns rules in a particular order, so make a lookup table
+	lookup := make(map[string]api.CreateRuleReq, len(rules))
+	for _, rule := range rules {
+		lookup[rule.Id] = rule
+	}
+	// Now check each returned rule for existence and matching fields
+	for _, rule := range resp.Rules {
+		if expectedRule, ok := lookup[rule.Id]; ok {
+			if !ruleMatches(expectedRule, *rule) {
 				return false
 			}
+		} else {
+			return false
+		}
+	}
+	return true
+}
 
-			// Note: we're ignoring type conversion, as asserting that would require us
-			// to replicate the mapping logic in tests.
-			// Instead, as we add ReadRule, and ListRules etc, methods, we can use this
-			// testing approach to write meaningful assertions.
-			return len(r.Rule.Conditions) == len(reqs.rules[0].Conditions) &&
-				r.Rule.Id == reqs.rules[0].Id &&
-				r.Rule.Name == reqs.rules[0].Name &&
-				r.Rule.ProjectId == reqs.rules[0].ProjectId
-		},
-		gopter.CombineGens(createProjectReqGen, createRuleReqGen).Map(func(vals []interface{}) projectAndRuleReq {
-			p := vals[0].(api.CreateProjectReq)
-			r := vals[1].(api.CreateRuleReq)
-			r.ProjectId = p.Id
-			return projectAndRuleReq{
-				CreateProjectReq: p,
-				rules:            []api.CreateRuleReq{r},
-			}
-		}),
-	))
-	properties.Property("ensures IDs are unique", prop.ForAll(
-		func(reqs projectAndRuleReq) bool {
-			defer testDB.Flush(t)
-			_, err := cl.CreateProject(ctx, &reqs.CreateProjectReq)
-			if err != nil {
-				t.Error(err.Error())
-				return false // bad run
-			}
+func createRules(ctx context.Context, cl api.ProjectsClient, rules []api.CreateRuleReq) ([]api.CreateRuleReq, error) {
+	for i, rule := range rules {
+		// make the id unique!
+		rules[i].Id = fmt.Sprintf("%s-index-%d", rules[i].Id, i)
```

Comment:
As soon as I introduced multiple rule generation, I ran into the issue that gopter does not make any claims on uniqueness. This patch handles that.

Reply:
I think it would be neat to fix that in the generator definiton: feels like it could be [part of the `Map` call above](https://github.com/chef/automate/pull/712/files#diff-66769720a1732f29e7206a24759eea52R584).

Reply:
Definitely; good catch!